### ISSUE TITLE
Describe templating rules

### ISSUE THREAD

**Issue by rin-st** [scaffold-eth/create-eth#132]
- [ ] objects, replace
- [ ] objects, merge
- [ ] arrays, replace
- [ ] arrays, merge
- [ ] strings
- [ ] other primitives?

https://github.com/scaffold-eth/create-eth/pull/131#discussion_r1789926986

**Comment by rin-st** [comment]
For replacing objects/arrays logic is simple: just use stringified variable.

As Shiv mentioned [here](https://github.com/scaffold-eth/create-eth/pull/129#discussion_r1787861953), we can use spread operators for adding new array items and merging oblects, examples.

| | Template | Args | Result |
|--------|--------|--------|--------|
| object, replace | const replacedObj = ${JSON.stringify(replacedObj[0])} | { key1: "Replaced", key2: "Object" } | const replacedObj = { key1: "Replaced", key2: "Object" } |
| array, replace | const replacedArr = ${JSON.stringify(replacedArr[0])} | ["Replaced", "Array"] | const replacedArr = ["Replaced", "Array"] |
| object, add new entries | const mergedObj = {key1: 'value1', key2: 'value2', ...${JSON.stringify(objToMerge[0])}} | { key3: "Merged", key4: "Object" } | const mergedObj = { key1: "value1", key2: "value2", ...{ key3: "Merged", key4: "Object" } }; |
| array, add new items | const arrWithAdditionalItems = ['a', 'b', ...${JSON.stringify(arrayToSpread[0])}] | ["Spread", "This"] | const arrWithAdditionalItems = ["a", "b", ...["Spread", "This"]] |
| BigInt | const bigInt = ${someBigInt[0]} | const someBigInt = 123n | const bigInt = 123 |

Notes: 
- when BigInt is used inside object/array variable, we cannot JSON.stringify that variable by default. Need to add BigInt stringifyer to the code. Most of the time [this](https://github.com/GoogleChromeLabs/jsbi/issues/30#issuecomment-1006086291) is used. But we need 
- if we use variable of type BigInt and then use it in code, it loses `n` at the end, so it converts to regular number. Not sure what to do with it, add `n` when handling that bigInt? Same problem with stringified bigints in objects/arrays

---
Regarding removing spread operator by linter, I found [this](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-useless-spread.md) but we'll need to install a plugin just to enable one rule. Didn't test it yet. I'll try test it to find another solutions next days

**Comment by carletex** [comment]
This is great, thanks @rin-st 

Are there instances when using template strings is required, right? e.g. When it's JSX code?

**Comment by technophile-04** [comment]
Thanks RInat!

> Are there instances when using template strings is required, right? e.g. When it's JSX code?

Yeah or even while providing code snippets like we do in `ScaffoldEthProvider.template.mjs` with [`providerSetup`](https://github.com/scaffold-eth/create-eth-extensions/blob/7b125cafc6859977ca280ab54aeebc35838c7174/extension/packages/nextjs/components/ScaffoldEthAppWithProviders.tsx.args.mjs#L2). 


> - when BigInt is used inside object/array variable, we cannot JSON.stringify that variable by default. Need to add BigInt stringifyer to the code. Most of the time https://github.com/GoogleChromeLabs/jsbi/issues/30#issuecomment-1006086291 is used. But we need
> 
>  - if we use variable of type BigInt and then use it in code, it loses n at the end, so it converts to regular number. Not sure what to do with it, add n when handling that bigInt? Same problem with stringified bigints in objects/arrays

Ohhh yeah its PITA :/ 

Also if we use point 1 hack while stringifying it will cause issue right?  

Lets suppose some `.args.mjs` file looks like this: 

```js
const test = {
   big: 134n
}
```

And the final file requirement type asserts that big needs to `bigInt` and not the string. 

So using point 1 hack won't be able to work right?  I think that what you mentioned in point 2? 

Yeah lol not sure what's the best solution will be, but yeah we can always thinking about this case when in future objects requires to have a property `bigint`.

**Comment by rin-st** [comment]
> And the final file requirement type asserts that big needs to bigInt and not the string.

Yes, I meant that

> So using point 1 hack won't be able to work right? I think that what you mentioned in point 2?

Yep. In the table last row bigint became a number, so if bigint is long it loses precision. So if dev wants to make it bigint in the code, he need to use for example `const bigInt = BigInt(${someBigInt[0]})`, we can add it to the rules too

```
const test = {
   big: 134n
}
```

> So using point 1 hack won't be able to work right? I think that what you mentioned in point 2?

For that case both 1 and 2 doesn't work. 2 works only for simple bigint variable. We can try to implement our function like in [mdn docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt#use_within_json) and mention it in rules too

**Comment by technophile-04** [comment]
>  For that case both 1 and 2 doesn't work. 2 works only for simple bigint variable. We can try to implement our function like in [mdn docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt#use_within_json) and mention it in rules too

Yeah :(, but yeah I think it will be rare case but lets keep it mind and lol we can alway fallback to hacky solutuion suggesting people to wrap whole object inside in template literal. So that we don't loose `n`. But I think this will be very rare case though

**Comment by rin-st** [comment]
I found a possible solution, not very good but ok. And for `...` too. Hopefully I will create PR later today
